### PR TITLE
Gh 955 federated gaffer store

### PR DESCRIPTION
**Pull request by GCHQDeveloper404**

### CODE REVIEW

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class FederatedStore extends Store {
```
Comment (t616178) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java]: Could you add some javadoc to this class (and the other new classes) please?

File: core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/io/InputImplTest.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2016 Crown Copyright
```
Comment (GCHQDeveloper404) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/io/InputImplTest.java]: Nope, I didn't create that class. The date remains from when it was created correct?

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java [hunk 3]
```diff
@@ -451,21 +451,21 @@ public void validateSchemas() {
             validateSchema(validationResult, getSchema().getVertexSerialiser());
 
             getSchema().getTypes().entrySet().forEach(entrySet ->
-                    validateSchema(validationResult, entrySet.getValue().getSerialiser()));
+                                                              validateSchema(validationResult, entrySet.getValue().getSerialiser()));
```
Comment (GCHQDeveloper404) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java]: I've been using the zip that was given to me from @p013570 unless mine has drifted away.
I'll check, against what you have.

File: core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/io/InputImplTest.java (hunk 2 above)
Comment (p013570) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/io/InputImplTest.java]: You can change it to 2016-2017

Comment (GCHQDeveloper404) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/io/InputImplTest.java]: I'll add 2016-2017 to my template.

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    public void initialise(final String graphId, final Schema unused, final StoreProperties properties) throws StoreException {
+        super.initialise(graphId, new Schema(), properties);
+        loadGraphs();
+    }
+
+    public void initialise(final String graphId, final StoreProperties properties) throws StoreException {
+        this.initialise(graphId, null, properties);
+    }
+
+    @Override
```
Comment (m55624) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java]: This isnt necessary

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+public class FederatedStore extends Store {
+    public static final String GAFFER_FEDERATEDSTORE = "gaffer.federatedstore.";
+    public static final String PROPERTIES = "properties";
+    public static final String SCHEMA = "schema";
+    public static final String SCHEMA_DEL_REGEX = Pattern.quote(",");
+    public static final String KEY_DEL_REGEX = Pattern.quote(".");
+    public static final String USER_IS_ATTEMPTING_TO_OVERWRITE_A_GRAPH_WITHIN_FEDERATED_STORE_GRAPH_ID_S = "User is attempting to overwrite a graph within FederatedStore. GraphId: %s";
+    public static final String GRAPH_WAS_NOT_ABLE_TO_BE_CREATED_WITH_THE_USER_SUPPLIED_PROPERTIES_GRAPH_ID_S = "Graph was not able to be created with the user supplied properties. GraphId: %s";
+    private final Map<String, Graph> graphs = Maps.newHashMap();
+    private Schema schema = new Schema();
```
Comment (m55624) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java]: Could we possibly rename this to federatedSchema and update the 'get' method.  It is slightly confusing having a getSchema here and a getSchema in the Store when they actually return different things.

File: store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/FederatedOperationHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import uk.gov.gchq.gaffer.operation.OperationChain;
+import uk.gov.gchq.gaffer.store.Context;
+import uk.gov.gchq.gaffer.store.Store;
+import uk.gov.gchq.gaffer.store.schema.Schema;
+import uk.gov.gchq.gaffer.user.User;
+import java.util.LinkedHashSet;
+
+public class FederatedOperationHandlerTest {
+    public static final String TEST_USER = "testUser";
+    protected User user;
```
Comment (m55624) [inline store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/FederatedOperationHandlerTest.java]: Can this be private?

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java [hunk 7]
```diff
@@ -456,16 +456,16 @@ public void validateSchemas() {
 
         if (!validationResult.isValid()) {
             throw new SchemaException("Schema is not valid. "
-                    + validationResult.getErrorString());
+                                              + validationResult.getErrorString());
```
Comment (GCHQDeveloper404) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java]: My formatting settings drifted away 😕

File: doc/src/main/java/uk/gov/gchq/gaffer/doc/dev/walkthrough/FederatedStoreWalkThrough.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+import uk.gov.gchq.gaffer.store.StoreProperties;
+import uk.gov.gchq.gaffer.store.schema.Schema;
+import uk.gov.gchq.gaffer.user.User;
+
+public class FederatedStoreWalkThrough extends DevWalkthrough {
+    public FederatedStoreWalkThrough() {
+        super("FederatedStore", "RoadAndRoadUseWithTimesAndCardinalities");
+    }
+
+    public CloseableIterable<? extends Element> run() throws Exception {
```
Comment (GCHQDeveloper404) [inline doc/src/main/java/uk/gov/gchq/gaffer/doc/dev/walkthrough/FederatedStoreWalkThrough.java]: Non-operatational code I went for cleaner code for beginners to read.

File: doc/src/main/java/uk/gov/gchq/gaffer/doc/walkthrough/AbstractWalkthrough.java [hunk 9]
```diff
@@ -56,7 +56,7 @@ public AbstractWalkthrough(final String header,
         storePropertiesLocation = "/mockaccumulostore.properties";
     }
 
-    public abstract Object run() throws OperationException, IOException;
+    public abstract Object run() throws Exception;
```
Comment (GCHQDeveloper404) [inline doc/src/main/java/uk/gov/gchq/gaffer/doc/walkthrough/AbstractWalkthrough.java]: Non-operatational code I went for cleaner code for beginners to read.

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java (hunk 5 above)
Comment (GCHQDeveloper404) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java]: Good spot!, this is actually hiding the private super.schema I've made the super protected and deleted this schema.
Take a cookie 🍪.

File: store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+    }
+
+    public void add(final Collection<Graph> graphs) {
+        graphs.forEach(this::_add);
+        updateMergedGraphConfig();
+    }
+
+    private void _add(final Graph graph) {
+        if (graphs.containsKey(graph.getGraphId())) {
+            throw new IllegalStateException((String.format(USER_IS_ATTEMPTING_TO_OVERWRITE_A_GRAPH_WITHIN_FEDERATED_STORE_GRAPH_ID_S, graph.getGraphId())));
```
Comment (GCHQDeveloper404) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java]: nice 👍

Comment (GCHQDeveloper404) [inline store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/FederatedStore.java]: nice 1 👍

File: doc/src/main/java/uk/gov/gchq/gaffer/doc/walkthrough/AbstractWalkthrough.java (hunk 9 above)
Comment (GCHQDeveloper404) [inline doc/src/main/java/uk/gov/gchq/gaffer/doc/walkthrough/AbstractWalkthrough.java]: Non-operational code, cleaner to read for beginners less noise on screen.

File: doc/src/main/java/uk/gov/gchq/gaffer/doc/walkthrough/WalkthroughStrSubstitutor.java [hunk 11]
```diff
@@ -174,7 +173,7 @@ public static void validateSubstitution(final String walkthrough, final String..
         if (null != example) {
             try {
                 example.run();
-            } catch (final OperationException | IOException e) {
+            } catch (final Exception e) {
```
Comment (GCHQDeveloper404) [inline doc/src/main/java/uk/gov/gchq/gaffer/doc/walkthrough/WalkthroughStrSubstitutor.java]: Non-operational code, cleaner to read for beginners less noise on screen.

File: core/operation/src/main/java/uk/gov/gchq/gaffer/operation/OperationChain.java [hunk 12]
```diff
@@ -116,6 +116,19 @@ public void close() throws IOException {
         }
     }
 
+    @Override
+    public boolean equals(final Object obj) {
```
Comment (GCHQDeveloper404) [inline core/operation/src/main/java/uk/gov/gchq/gaffer/operation/OperationChain.java]: This has been done, but GitHub isn't displaying the change here.

**Comment by GCHQDeveloper404** [comment]
Going to add changes such as graph library.

**Pull request closed without merging**